docs/options.md
# Command line options

**Content**
* [Input and output](#input-and-output)
  * [--pattern](#--pattern) &nbsp; [--output](#--output) &nbsp; [--output-blend](#--output-blend) &nbsp;
[--temp-dir](#--temp-dir) &nbsp;  [--frames](#--frames) &nbsp;  [--quality](#--quality)
* [Algorithm](#algorithm)
  * [--mode](#--mode) &nbsp; [--threshold](#--threshold) &nbsp; [--outlier](#--outlier) &nbsp;
[--background](#--background) &nbsp; [--weights](#--weights) &nbsp; [--fade](#--fade)
* [Video creation](#video-creation)
  * [--video-in](#--video-in) &nbsp; [--video-out](#--video-out)
* [Camera shake reduction](#camera-shake-reduction)
  * [--shake](#--shake) &nbsp; [--shake-anchors](#--shake-anchors)
* [Performance](#performance)
  * [--threads](#--threads) &nbsp; [--video-threads](#--video-threads) &nbsp; [--shake-threads](#--shake-threads) &nbsp; [--sample](#--sample) &nbsp; [--compression](#--compression) &nbsp; [--slice](#--slice)
* [Misc](#misc)
  * [--debug](#--debug) &nbsp; [--wait](#--wait)

## Input and output

#### `--pattern`

_Required._
Search pattern for input files (glob-style).

_**Note:**_ The pattern **MUST be quoted on Unix** systems.

Examples:
```
--pattern "path/to/*.jpg"
--pattern "image-*.jpg"
```
The files found are used in lexical order.

#### `--output`

_Required._ Output file path. File format is determined from the extension. Supported formats are JPG, PNG and TIF. 
For a list of formats potentially supported, but not tested, see crate [`image`](https://docs.rs/image/).

Examples:
```
--output path/to/out.png
```

#### `--output-blend`

_Optional, used with `--mode outlier` only._

Output path for the greyscale image showing the algorithm's outlier detections. See [`--output`](#--output) for details.

_Default:_ No output of outlier image.

#### `--temp-dir`

_Optional, used with `--mode outlier` only._

Temporary directory for storing time slice files. 
Files are delected after processing, while the directory is not. 
If the directory does not exist, but the parent directory exists, it is created.
If the parent directory does not exist, the program exits with an error.

_Default:_ `<system-temp>/chrono-photo/`

#### `--frames`

_Optional._ A range of frames in the format `start/end/step`, where `end` is exclusive.

The `.` character can be used as placeholder.

Examples:
```
--frames ././2     -> Use every second frame (of all frames; no start or end given)
--frames ./20/.    -> Use the first 20 frames (zero-based indexing, end exclusive)
```

_Default:_ Use all images found for [`--pattern`](#--pattern).

#### `--quality`

_Optional._ Output quality for JPEG images, in percent (1 - 100).

Examples:
```
--quality 100
```

_Default:_ `95`

## Algorithm

#### `--mode`

_Optional._ Foreground pixel detection mode. One of `outlier|darker|lighter`.

_Default:_ `outlier`

#### `--threshold`

_Optional._ Outlier distance to median threshold and blend distance in format `(abs|rel)/min[/max]`.
Colors closer to the median than `min` are considered background.
Colors between `min` and `max` are blended over the background linearly.
Colors more distant to the median then `max` are blended over the background with 100% (replace, but see [--fade](#--fade)).

An **absolute** threshold relates to the total color range, e.g. fractions of 255 for 8 bit per channel images.
E.g. `abs/0.1/0.2` for 8 bits images blends for distances between 25.5 and 51 (0.1 * 255, 0.2 * 255).

A **relative** threshold relates to the inter-quartile range (IQR, difference between upper and lower quartile)
of the pixel's colors (separately for each channel / dimension).
E.g. `rel/2/4` blends for distances between twice the IQR to four times the IQR.

Examples:
```
--threshold abs/0.05/0.2
--threshold rel/3.0/5.0
```

_Default:_ `abs/0.05/0.2`

#### `--outlier`

_Optional, used with `--mode outlier` only._

Outlier selection mode. Determines how selection between multiple outliers is performed.

This parameter is particularly useful when several movement trails overlap,
or when a moving object is relatively slow compared to the recording frame rate, and thus overlaps itself. 
For non-overlapping trails, the default `extreme` should be the best.

* `extreme`: use the most extreme outlier (the default value).
* `average`: use the average of all outliers.
* `forward`, `backward`: progressively blends all outliers over the background, starting with the first or last, respectively.
* `first`: use the first outlier found.
* `last`: use the last outlier found.

_Default:_ `extreme`

#### `--background`

_Optional, used with `--mode outlier` only._

Background color selection mode. Determines how selection between non-outlier colors is performed.

* `random`: Use a randomly selected pixel value, selected among all non-outlier images. The default value, but may result in a noisy image.
* `first`: Use the pixel value from the first non-outlier image.
* `average`: Use the average pixel value of all non-outlier images. Can be used for motion blur, but may result in banding for low contrast backgrounds.
* `median`: Use the median pixel value of all images (including outliers!). May result in banding for low contrast backgrounds.

_Default:_ `random`

#### `--weights`

_Optional._ Color channel weights for outlier detection in format `r g b a`.
Must be four numeric values (typically 0 - 1), irrespective of the actual number of color channels in the input images.

Examples:
```
--weights 0 0 1 0              -> Outlier analysis is based only on the blue channel
--weights 1.0 0.5 0.5 0.0      -> Outlier analysis is based on more weight on red than on blue and green
```

_Default:_ `1 1 1 1`

#### `--fade`

_Optional._ Allows for fading outlier blending over frames. Format `(clamp|repeat)/(abs/rel)/f,v/f,v[/f,v...]`

Parts between `/` are:
1. Fading mode: `clamp` or `repeat`. Specifies how frames outside the given fade transition are treated.
1. `abs` or `rel`: specifies where the frames given in the `frame,value` parts relate to. 
`abs` is relative to the first frame in the entire sequence (forward). 
`rel` is relative to the last actually processed frame (backward).
See the examples.
1. At least two blend value pairs in the format `frame,value` (without additional spaces!).

**Must be in strictly increasing frame order!**

Between the specified frames, values are interpolated linearly. Values should be between 0 and 1.

Examples:
```
--fade repeat/rel/0,1/9,0   -> Repeated, fading-out movement trails over 10 'instances' of a moving object
                               Would move when use for video output
```

_Default:_ No fading.

## Video creation

For video creation, at least one of the two options `--video-in` and `--video-out` must be provided.

For video output, the frame counter is appended to the file name of the path provided by option `--output`.
E.g. `out.jpg` becomes `out-00000.jpg`, `out-00001.jpg`, etc.

#### `--video-in`

_Optional._ Frame range of input images per video frame, relative to the current video frame. 
Format `start/end/step`, where `end` is exclusive. Use `.` as placeholder.

If `start` and `end` are given, it results in a 'moving window' over all images. 
If `start` or `end` are placeholders, the frames range from the very start, 
or to the very end (of the entire image sequence), respectively.

Examples:
```
--video-in 0/25/.     -> Each video frame will contain 25 frames. 
                         Results in a 'moving trails'.
--video-in ./1/.      -> Each video frame will contain all images up to the current frame.
                         Results in a 'growing trail'.
--video-in 0/50/5     -> Each video frame will contain every 5th image of the given range.
                         Results in the impression of the moving object 'following itself' multiple times 
                         in a certain distance. 
                         Requires start and end to be specified in order to work as intended!
```

Everything here refer to the frames left _after_ selection through option [`--frames`](#--frames)!

_Default:_ No video output, or `././.` if `--video-out` is specified.

#### `--video-out`

_Optional._ Frame range of the video. in format `start/end/step`, where `end` is exclusive. Use `.` as placeholder.

The actual frame range will most likely be larger, as it is automatically extended
to cover the per-frame offsets given by option `--video-in`.

In most cases, it is not necessary to set this option and just use `--video-in`.

Everything here refer to the frames left _after_ selection through option [`--frames`](#--frames)!

_Default:_ No video output, or `././.` if `--video-in` is specified.

## Camera shake reduction

To enable camera shake reduction, both of the following options must be supplied.
By default, no camera shake reduction is applied.

If camera shake is detected, images are cropped by the amount of shake for correction.
Thus, the output image will be slightly smaller than the input images.

#### `--shake`

_Optional._ Shake anchor radius and search radius in format `anchor-radius/search-radius`.

Example:
```
--shake 10/5
```
_Default:_ No camera shake reduction.

#### `--shake-anchors`

_Optional._ Pixel coordinates of shake detection anchors in the _first_ image. 
Format `x1/y1 [x2/y2 ...]`.

Anchors are optimally placed at positions with high contrast in both directions (x, y),
like a dark corner on light background.
Also, anchors should not be occluded by a moving object in any image of the sequence.

Use an image editing software to get the exact pixel coordinates.
Origin is the top-left corner of the image.

Example:
```
--shake-anchors 1234/789 2345/890
```

_Default:_ No camera shake reduction.

## Performance

#### `--threads`

_Optional._

Number of threads to use for parallel processing.

_Default:_ Number of processors.

#### `--video-threads`

_Optional._

Number of threads to use for parallel processing for video / image sequence creation. 
Entire frames are processed in parallel. 
It may be necessary to limit the number of video threads if memory usage is too high.

_Default:_ Number of processors.

#### `--shake-threads`

_Optional._

Number of threads to use for parallel camera shake analysis. 
Entire frames are processed in parallel. 
It may be necessary to limit the number of shake threads if memory usage is too high.

_Default:_ Number of processors.

#### `--sample`

_Optional, used with `--mode outlier` only._

Specifies a sample count to reduce the number of images used for calculation of per-pixel median
and (if required) quartiles.

This option is particularly useful to speed up calculations when processing large numbers of images (thousands).

_Default:_ No sampling, use all images.

#### `--compression`

_Optional, used with `--mode outlier` only._

Compression method and level for temporary time slice files. 
Format `(gzip|zlib|deflate)[/<level>]`. 
Levels range from 0 (no compression) to 9 (slowest).

_Default:_ `gzip/6`

#### `--slice`

_Optional, used with `--mode outlier` only._

Time-slicing in the format `(rows|pixels|count)/<number>`.

_Default:_ `rows/4` (should be sufficient for most scenarios)

Holding a large number of high resolution images in memory at the same time is not feasible. 

Therefore, before actual processing, the time-stack of images with (x, y) coordinates
is converted into a number of temporary files, each containing data in (x, t) coordinates.
For example, the first temporary file contains the first row of pixels from each image.

Using these temporary files, all images can be processes row by row, without overloading memory, as explained above.

Actually, the above description is a simplification. Option `--slice` provides control
over how much data from each image goes into each temporary file. The option accepts different forms.

Examples:
```
--slice rows/4         -> Writes 4 rows of each image into each time slice.
--slice pixels/1000    -> Writes 1000 pixels of each image into each time slice.
--slice count/100      -> Internally determines the amount of data written, in order to create 
                          a total of 100 time slices.
```

## Misc

#### `--debug`

_Optional._ Switch to print the parsed command line arguments for debugging.

#### `--wait`

_Optional._ Switch to keep the terminal open and wait for user key press.
Useful when running a `.bat`, `.sh` or `.chrono` file by double click,
to let the user check for errors.


src/cli.rs
//! Command-line interface for chrono-photo.
use crate::flist::FrameRange;
use crate::options::{BackgroundMode, Fade, OutlierSelectionMode, SelectionMode, Threshold};
use crate::shake::{ShakeAnchor, ShakeParams, ShakeReduction};
use crate::slicer::SliceLength;
use crate::streams::Compression;
use core::fmt;
use std::path::PathBuf;
use std::str::FromStr;
use structopt::StructOpt;

/// Command-line tool for combining images into a single chrono-photograph or chrono-video.
///
/// Use `chrono-photo -h`     for help, or
///     `chrono-photo --help` for more detailed help.
///
/// For more documentation and explanation of the algorithm, see the GitHub repository:
///      https://mlange-42.github.io/chrono-photo/
#[derive(StructOpt)]
#[structopt(verbatim_doc_comment)]
//#[structopt(name = "chrono-photo command line application")]
pub struct Cli {
    /// File search pattern. ** MUST be quoted on Unix systems! **
    #[structopt(short, long)]
    pattern: String,

    // /// Sets to 16 bit per color channel mode. Use for 16 bit TIFF files.
    // #[structopt(long, name = "16-bit")]
    // is_16bit: bool,
    /// Frames to be used from those matching pattern: `start/end/step`. Optional.
    /// For default values, use `.`, e.g. `././2`.
    #[structopt(short, long, value_name = "frames")]
    frames: Option<FrameRange>,

    /// Video input frames. Frames to be used per video frame: `start/end/step`. Optional.
    #[structopt(long, name = "video-in", value_name = "frames")]
    video_in: Option<FrameRange>,

    /// Video output frames. Range and step width of video output frames: `start/end/step`. Optional.
    #[structopt(long, name = "video-out", value_name = "frames")]
    video_out: Option<FrameRange>,

    /// Path to output file
    #[structopt(short, long, value_name = "path")]
    output: String,

    /// Temp directory. Used with `--mode outlier` only. Optional, default system temp directory.
    #[structopt(long, name = "temp-dir", value_name = "path")]
    temp_dir: Option<String>,

    /// Path of output image showing which pixels are outliers (blend value).
    /// Used with `--mode outlier` only.
    #[structopt(long, name = "output-blend", value_name = "path")]
    output_blend: Option<String>,

    /// Pixel selection mode (lighter|darker|outlier). Optional, default 'outlier'.
    #[structopt(short, long)]
    mode: Option<SelectionMode>,

    /// Outlier threshold mode (abs|rel)/<lower>[/<upper>]. Optional, default 'abs/0.05/0.2'.
    /// Used with `--mode outlier` only.
    #[structopt(short, long, value_name = "thresh")]
    threshold: Option<Threshold>,

    /// Background pixel selection mode (first|random|average|median). Optional, default 'random'.
    /// Used with `--mode outlier` only.
    #[structopt(short, long, value_name = "bg")]
    background: Option<BackgroundMode>,

    /// Outlier selection mode in case more than one outlier is found
    /// (first|last|extreme|average|forward|backward). Optional, default 'extreme'.
    /// Used with `--mode outlier` only.
    #[structopt(short = "l", long, value_name = "mode")]
    outlier: Option<OutlierSelectionMode>,

    /// Compression mode and level (0 to 9) for time slices (gzip|zlib|deflate)[/<level>].
    /// Used with `--mode outlier` only.
    /// Optional, default 'gzip/6'.
    #[structopt(short, long, value_name = "comp/lev")]
    compression: Option<Compression>,

    /// Output image quality for JPG files, in percent. Optional, default '95'.
    #[structopt(short, long)]
    quality: Option<u8>,

    /// Controls slicing to temp files (rows|pixels|count)/<number>.
    /// Used with `--mode outlier` only.
    /// Optional, default 'rows/4'.
    #[structopt(short, long)]
    slice: Option<SliceLength>,

    /// Restricts calculation of median and inter-quartile range to a sub-sample of input images.
    /// Use for large amounts of images to speed up calculations. Optional.
    /// Used with `--mode outlier` only.
    #[structopt(long)]
    sample: Option<usize>,

    /// Color channel weights (4 values: RGBA) for distance calculation. Optional, default '1 1 1 1'.
    #[structopt(long, number_of_values = 4, value_name = "w")]
    weights: Option<Vec<f32>>,

    /// Frame fading. Optional, default None. Format: (clamp|repeat)/(abs|rel)/(f1,v1)/(f2,v2)[/(f,v)...]
    #[structopt(long)]
    fade: Option<Fade>,

    /// Number of threads. Optional, default equal to number of processors.
    #[structopt(long, value_name = "num")]
    threads: Option<usize>,

    /// Number of threads for parallel video frame output. Optional, default equal to number of processors.
    /// Limiting this may be required if memory usage is too high.
    #[structopt(long, name = "video-threads", value_name = "num")]
    video_threads: Option<usize>,

    /// Number of threads for camera shake reduction. Optional, default equal to number of processors.
    /// Limiting this may be required if memory usage is too high.
    #[structopt(long, name = "shake-threads", value_name = "num")]
    shake_threads: Option<usize>,

    /// Camera shake reduction parameters. Optional, default none.
    /// Format: `anchor-radius/shake-radius`
    #[structopt(long, value_name = "r1/r2")]
    shake: Option<ShakeParams>,

    /// Camera shake reduction anchors. Optional, default none. Format: `x1/y1 [x2/y2 [...]]`
    #[structopt(long, name = "shake-anchors", value_name = "x/y")]
    shake_anchors: Option<Vec<ShakeAnchor>>,

    /// Prints debug information (i.e. parsed cmd parameters) before processing.
    #[structopt(long, short)]
    debug: bool,

    /// Keeps the terminal open after processing and waits for user key press.
    #[structopt(long, short)]
    wait: bool,
}

impl Cli {
    /// Parses this Cli into a [CliParsed](struct.CliParsed.html).
    pub fn parse(self) -> Result<CliParsed, ParseCliError> {
        let mut warings = Vec::new();
        if self.mode.is_some() && self.mode.as_ref().unwrap() != &SelectionMode::Outlier {
            if self.output_blend.is_some() {
                warings.push("--output-blend".to_string());
            }
            if self.threshold.is_some() {
                warings.push("--threshold".to_string());
            }
            if self.outlier.is_some() {
                warings.push("--outlier".to_string());
            }
            if self.background.is_some() {
                warings.push("--background".to_string());
            }
            if self.temp_dir.is_some() {
                warings.push("--temp-dir".to_string());
            }
            if self.sample.is_some() {
                warings.push("--sample".to_string());
            }
            if self.slice.is_some() {
                warings.push("--slice".to_string());
            }
            if self.compression.is_some() {
                warings.push("--compression".to_string());
            }
        }
        if self.shake.is_some() != self.shake_anchors.is_some() {
            return Err(ParseCliError(
                "Provide both options or none: `--shake` and `--shake-anchors`".to_string(),
            ));
        }

        let mut weights = [1.0; 4];
        if let Some(w) = &self.weights {
            for (i, v) in w.iter().enumerate() {
                weights[i] = *v;
            }
        }

        let shake_params = self.shake;
        let shake_anchors = self.shake_anchors;
        let out = CliParsed {
            pattern: self.pattern,
            // is_16bit: self.is_16bit,
            temp_dir: self.temp_dir.map(|d| PathBuf::from(d)),
            output: PathBuf::from(&self.output),
            output_blend: match self.output_blend {
                Some(out) => Some(PathBuf::from(out)),
                None => None,
            },
            mode: self.mode.unwrap_or(SelectionMode::Outlier),
            threshold: self.threshold.unwrap_or(Threshold::abs(0.05, 0.2)),
            background: self.background.unwrap_or(BackgroundMode::Random),
            outlier: self.outlier.unwrap_or(OutlierSelectionMode::Extreme),
            compression: self.compression.unwrap_or(Compression::GZip(6)),
            quality: match self.quality {
                Some(q) => {
                    if q <= 100 && q > 0 {
                        q
                    } else {
                        return Err(ParseCliError(format!(
                            "Expected 0 < qualtiy <= 100. Got value {}",
                            q
                        )));
                    }
                }
                None => 95,
            },
            frames: self.frames,
            video_in: self.video_in,
            video_out: self.video_out,
            slice: self.slice.unwrap_or(SliceLength::Rows(4)),
            sample: self.sample,
            weights,
            fade: self.fade.unwrap_or(Fade::none()),
            threads: self.threads,
            video_threads: self.video_threads,
            shake_threads: self.shake_threads,
            shake_reduction: shake_params.and_then(|shake| {
                shake_anchors.and_then(|anchors| {
                    Some(ShakeReduction::new(
                        anchors.iter().map(|a| a.anchor()).collect(),
                        shake.anchor_radius(),
                        shake.search_radius(),
                    ))
                })
            }),
            debug: self.debug,
            wait: self.wait,
        };

        if !warings.is_empty() {
            println!("WARNING! The following options are not used, as they are required only for `--mode outlier`:", );
            for w in warings {
                println!("{}", w);
            }
            println!();
        }

        out.validate()
    }
}

impl FromStr for Cli {
    type Err = ParseCliError;

    /// Parses a string into a Cli.
    fn from_str(str: &str) -> Result<Self, Self::Err> {
        let quote_parts: Vec<_> = str.split('"').collect();
        let mut args: Vec<String> = vec![];
        for (i, part) in quote_parts.iter().enumerate() {
            let part = part.trim();
            if i % 2 == 0 {
                args.extend(
                    part.split(' ')
                        .map(|s| s.trim().to_string())
                        .filter(|s| !s.is_empty()),
                );
            } else {
                args.push(part.to_string());
            }
        }
        Ok(Cli::from_iter(args.iter()))
    }
}

/// Parsed command line arguments.
#[allow(dead_code)]
#[derive(Debug)]
pub struct CliParsed {
    /// File search pattern
    pub pattern: String,
    // /// Sets to 16 bit per color channel mode. Use for 16 bit TIFF files.
    //pub is_16bit: bool,
    /// Frames to be used from those matching pattern: `start/end/step`. Optional.
    /// For default values, use `.`, e.g. `././step`.
    pub frames: Option<FrameRange>,
    /// Video input frames. Frames to be used per video frame: `start/end/step`. Optional.
    pub video_in: Option<FrameRange>,
    /// Video output frames. Range and step width of video output frames: `start/end/step`. Optional.
    pub video_out: Option<FrameRange>,
    /// Temp directory. Uses system temp directory if `None`.
    pub temp_dir: Option<PathBuf>,
    /// Path of the final output image.
    pub output: PathBuf,
    /// Path of output image showing which pixels are outliers (blend value).
    pub output_blend: Option<PathBuf>,
    /// Pixel selection mode.
    pub mode: SelectionMode,
    /// Outlier threshold mode.
    pub threshold: Threshold,
    /// Outlier selection mode in case more than one outlier is found.
    pub outlier: OutlierSelectionMode,
    /// Background pixel selection mode.
    pub background: BackgroundMode,
    /// Compression mode for time slices.
    pub compression: Compression,
    /// Output image quality for JPG files, in percent.
    pub quality: u8,
    /// Controls slicing to temp files (rows|pixels|count)/<number>. Optional, default 'rows/1'
    pub slice: SliceLength,
    /// Restricts calculation of median and inter-quartile range to a sub-sample of input images. Use for large amounts of images to speed up calculations. Optional.
    pub sample: Option<usize>,
    /// Color channel weights for distance calculation
    pub weights: [f32; 4],
    /// Frame fading. Optional, default None.
    pub fade: Fade,
    /// Number of threads. Optional, default equal to number of processors.
    pub threads: Option<usize>,
    /// Number of threads for parallel video frame output. Optional, default equal to number of processors.
    pub video_threads: Option<usize>,
    /// Number of threads for camera shake reduction. Optional, default equal to number of processors.
    pub shake_threads: Option<usize>,
    /// Shake reduction
    pub shake_reduction: Option<ShakeReduction>,
    /// Print debug information (i.e. parsed cmd parameters).
    pub debug: bool,

    /// Keep the terminal and UI open after processing and wait for user key press.
    pub wait: bool,
}

impl CliParsed {
    /// Check for validity
    pub fn validate(self) -> Result<Self, ParseCliError> {
        Ok(self)
    }
}

/// Error type for failed parsing command line argument.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct ParseCliError(String);

impl fmt::Display for ParseCliError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        self.0.fmt(f)
    }
}

#[cfg(test)]
mod test {
    use crate::cli::Cli;

    #[test]
    fn args_from_string() {
        let str = "chrono-photo --pattern \"test_data/generated/*.jpg\" --output test_data/temp --weights 0 1 1 0";
        let cli: Cli = str.parse().unwrap();
        let parsed = cli.parse().unwrap();

        //println!("{:#?}", parsed);
    }
}
